### PR TITLE
Refactor GenerateCheckBox

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #5981 and #5983 

cc @NTaylorMullen @dougbu

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs [hunk 1]
```diff
@@ -81,11 +81,22 @@ public static TheoryData MultiAttributeCheckBoxData
             string expectedAttributeString)
         {
             // Arrange
-            var originalContent = "original content";
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: The changes to this file are very hard to follow. I _think_ you're adding additional checks of `PreElement` and so on but don't know why.

Please revert as much as possible that's not directly related to the `InputTagHelper` changes i.e. don't do much more than add `expectedAttributes` (perhaps pass in the `expectedAttributes`) and split `expectedPostElement` from `expectedContent`.

Alternatively, let's sit down tomorrow and discuss why the new code is an improvement.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 2]
```diff
@@ -279,54 +279,30 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
...
+                    ViewContext.FormContext.EndOfFormContent.Add(hiddenForCheckboxTag);
+                }
+                else
+                {
+                    output.PostElement.AppendHtml(hiddenForCheckboxTag);
                 }
             }
 
-            var checkBoxTag = Generator.GenerateCheckBox(
+            var tagBuilder = Generator.GenerateCheckBox(
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Nah, this is nice for debugging.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Useless to have two variables with the same value just for debugging. That's why I mentioned the caller.

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Different styles is all. I don't like to build methods based off of how a caller treats their return values 😕

Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: All the other methods are like what @dougbu is suggesting so I will do that

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs (hunk 1 above)
Comment (jbagga) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: The original test has `PreContent` (originalContent) and `Content` (expectedAttributeString) https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs#L86 so I added those. And `PostElement` (the hidden element) is now added for all checkbox type input tag helpers so I added a check for that. I can remove `PostContent` though.

**Comment by jbagga** [comment]
@dougbu Updated

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs [hunk 3]
```diff
@@ -108,10 +108,9 @@ public static TheoryData MultiAttributeCheckBoxData
             await tagHelper.ProcessAsync(context, output);
 
             // Assert
-            Assert.Empty(output.Attributes); // Moved to Content and cleared
-            Assert.Equal(expectedContent, HtmlContentUtilities.HtmlContentToString(output.Content));
-            Assert.Equal(TagMode.SelfClosing, output.TagMode);
-            Assert.Null(output.TagName); // Cleared
+            Assert.NotNull(output.PostElement);
+            Assert.Equal(originalContent, HtmlContentUtilities.HtmlContentToString(output.Content));
+            Assert.Equal(expectedContent, HtmlContentUtilities.HtmlContentToString(output));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: Will likely be cleaner to change the other updated tests similarly. But, I'm not going to hold you to that.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs [hunk 4]
```diff
@@ -495,47 +511,22 @@ public void Process_GeneratesFormattedOutput(string specifiedType, string expect
             output.Content.AppendHtml(originalContent);
             output.PostContent.AppendHtml(expectedPostContent);
 
-            var htmlGenerator = new Mock<IHtmlGenerator>(MockBehavior.Strict);
-            var tagHelper = GetTagHelper(htmlGenerator.Object, model: false, propertyName: nameof(Model.IsACar));
+            var htmlGenerator = new TestableHtmlGenerator(new EmptyModelMetadataProvider());
+            var tagHelper = GetTagHelper(htmlGenerator, model: false, propertyName: nameof(Model.IsACar));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: Revert this change. The test is about verifying the exact calls into the generator.

**Comment by jbagga** [comment]
@dougbu Updated
